// optics-agent/dist/Agent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_REPORT_INTERVAL_MS = exports.MIN_REPORT_INTERVAL_MS = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }(); // This file contains the Agent class which is the public-facing API
// for this package.
//
// The Agent holds the configuration and all the in-memory state for
// the server.


var _Normalize = require('./Normalize');

var _Instrument = require('./Instrument');

var _Report = require('./Report');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var MIN_REPORT_INTERVAL_MS = exports.MIN_REPORT_INTERVAL_MS = 10 * 1000;
var DEFAULT_REPORT_INTERVAL_MS = exports.DEFAULT_REPORT_INTERVAL_MS = 60 * 1000;

var Agent = function () {
  function Agent(options) {
    var _this = this;

    _classCallCheck(this, Agent);

    // Public options. See README.md for descriptions.
    var _ref = options || {},
        apiKey = _ref.apiKey,
        debugFn = _ref.debugFn,
        normalizeVersion = _ref.normalizeVersion,
        normalizeQuery = _ref.normalizeQuery,
        endpointUrl = _ref.endpointUrl,
        proxyUrl = _ref.proxyUrl,
        reportIntervalMs = _ref.reportIntervalMs,
        printReports = _ref.printReports,
        reportTraces = _ref.reportTraces,
        reportVariables = _ref.reportVariables,
        shutdownGracefully = _ref.shutdownGracefully;
    // XXX We don't actually intend for these fields to be part of a public
    //     stable API. https://github.com/apollostack/optics-agent-js/issues/51


    this.apiKey = apiKey || process.env.OPTICS_API_KEY;
    // XXX actually use debugFn
    this.debugFn = debugFn || console.log; // eslint-disable-line no-console

    // Ensure we have an api key. If not, print and disable the agent.
    if (!this.apiKey) {
      this.debugFn('Optics agent disabled: no API key specified. ' + 'Set the `apiKey` option to `configureAgent` or `new Agent`, ' + 'or set the `OPTICS_API_KEY` environment variable.');
      this.disabled = true;
      return;
    }
    this.disabled = false;

    this.normalizeVersion = normalizeVersion || _Normalize.normalizeVersion;
    this.normalizeQuery = normalizeQuery || _Normalize.normalizeQuery;
    this.endpointUrl = endpointUrl || process.env.OPTICS_ENDPOINT_URL || 'https://optics-report.apollodata.com/';
    this.endpointUrl = this.endpointUrl.replace(/\/$/, '');
    this.proxyUrl = proxyUrl || process.env.HTTPS_PROXY;
    this.printReports = !!printReports;
    this.reportTraces = reportTraces !== false;
    this.reportVariables = reportVariables !== false;

    this.reportIntervalMs = reportIntervalMs || DEFAULT_REPORT_INTERVAL_MS;
    if (this.reportIntervalMs < MIN_REPORT_INTERVAL_MS) {
      this.debugFn('Optics: minimum reportInterval is ' + MIN_REPORT_INTERVAL_MS + '. Setting reportInterval to minimum.');
      this.reportIntervalMs = MIN_REPORT_INTERVAL_MS;
    }

    // Internal state.

    // Data we've collected so far this report period.
    this.pendingResults = {};
    // The wall clock time for the beginning of the current report period.
    this.reportStartTime = +new Date();
    // The HR clock time for the beginning of the current report
    // period. We record this so we can get an accurate duration for
    // the report even when the wall clock shifts or drifts.
    this.reportStartHrTime = process.hrtime();

    // Interval to send the reports. Per
    // https://github.com/apollostack/optics-agent-js/issues/4 we may
    // want to make this more complicated than just setInterval.
    // XXX there's no way to stop this interval (eg, for tests)
    this.reportTimer = setInterval(function () {
      return _this.sendStatsReport();
    }, this.reportIntervalMs);

    // Shutdown gracefully on SIGINT and process.exit()
    this.shutdownGracefully = shutdownGracefully !== false;
    if (this.shutdownGracefully) {
      process.on('exit', function () {
        return _this.stop();
      });
      process.on('SIGINT', function () {
        var isOnlyListener = process.listenerCount('SIGINT') === 1;
        _this.stop(function (err) {
          if (isOnlyListener) {
            if (err && !process.exitCode) {
              process.exitCode = 1;
            }
            process.exit();
          }
        });
      });
    }
  }

  _createClass(Agent, [{
    key: 'instrumentSchema',
    value: function instrumentSchema(schema) {
      if (this.disabled) {
        return schema;
      }
      this.schema = (0, _Instrument.instrumentSchema)(schema, this);
      (0, _Report.reportSchema)(this, schema);
      return this.schema;
    }
  }, {
    key: 'koaMiddleware',
    value: function koaMiddleware() {
      if (this.disabled) {
        return function (_ctx, next) {
          return next();
        };
      }

      return _Instrument.koaMiddleware;
    }
  }, {
    key: 'middleware',
    value: function middleware() {
      if (this.disabled) {
        return function (_req, _res, next) {
          next();
        };
      }
      return _Instrument.opticsMiddleware;
    }
  }, {
    key: 'instrumentHapiServer',
    value: function instrumentHapiServer(server) {
      if (this.disabled) {
        return;
      }
      (0, _Instrument.instrumentHapiServer)(server);
    }
  }, {
    key: 'context',
    value: function context(req) {
      if (this.disabled) {
        return {};
      }
      return (0, _Instrument.newContext)(req, this);
    }
  }, {
    key: 'stop',
    value: function stop(handler) {
      if (this.reportTimer) {
        clearInterval(this.reportTimer);
        this.reportTimer = false;
        this.sendStatsReport(handler);
      }
    }

    // XXX This is not part of the public API.
    //     https://github.com/apollostack/optics-agent-js/issues/51

  }, {
    key: 'sendStatsReport',
    value: function sendStatsReport(handler) {
      if (!this.schema) {
        this.debugFn('Optics agent: schema not instrumented. Make sure to call `instrumentSchema`.');
        return;
      }
      // copy current report state and reset pending state for the next
      // report.
      var reportData = this.pendingResults;
      var oldStartTime = this.reportStartTime;
      var durationHr = process.hrtime(this.reportStartHrTime);
      this.reportStartHrTime = process.hrtime();
      this.reportStartTime = +new Date();
      this.pendingResults = {};
      // actually send
      (0, _Report.sendStatsReport)(this, reportData, oldStartTime, this.reportStartTime, durationHr, handler);
    }
  }]);

  return Agent;
}();

exports.default = Agent;
