// optics-agent/dist/DefaultAgent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.context = exports.instrumentHapiServer = exports.koaMiddleware = exports.middleware = exports.instrumentSchema = exports.configureAgent = undefined;

var _Agent = require('./Agent');

var _Agent2 = _interopRequireDefault(_Agent);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var defaultAgent = null;

// This is used to pass options to the default Agent. Call it at most
// once before any other calls.
// This file contains the global state for default Agent object. Most
// users will only want one agent, and it can be convienient not to
// have to share an Agent instance between the various instrumentation
// call sites.

var configureAgent = exports.configureAgent = function configureAgent(options) {
  if (defaultAgent) {
    throw new Error('Optics Agent configuration already set. Make sure `configureAgent` is called only once and before any calls to `middleware`, `instrumentSchema`, or `instrumentHapiServer`.');
  }

  defaultAgent = new _Agent2.default(options);
};

var instrumentSchema = exports.instrumentSchema = function instrumentSchema(schema) {
  if (!defaultAgent) {
    defaultAgent = new _Agent2.default();
  }
  return defaultAgent.instrumentSchema(schema);
};

var middleware = exports.middleware = function middleware() {
  if (!defaultAgent) {
    defaultAgent = new _Agent2.default();
  }
  return defaultAgent.middleware();
};

var koaMiddleware = exports.koaMiddleware = function koaMiddleware() {
  if (!defaultAgent) {
    defaultAgent = new _Agent2.default();
  }
  return defaultAgent.koaMiddleware();
};

var instrumentHapiServer = exports.instrumentHapiServer = function instrumentHapiServer(server) {
  if (!defaultAgent) {
    defaultAgent = new _Agent2.default();
  }
  return defaultAgent.instrumentHapiServer(server);
};

var context = exports.context = function context(req) {
  if (!defaultAgent) {
    defaultAgent = new _Agent2.default();
  }
  return defaultAgent.context(req);
};
