// optics-agent/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Agent = exports.context = exports.instrumentHapiServer = exports.koaMiddleware = exports.middleware = exports.instrumentSchema = exports.configureAgent = undefined;

var _Agent = require('./Agent');

var _Agent2 = _interopRequireDefault(_Agent);

var _DefaultAgent = require('./DefaultAgent');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// export both as individual symbols and as a default object to
// support both of these syntaxes:
//   import OpticsAgent from 'optics-agent';
//   import { middleware, instrumentSchema } from 'optics-agent';
//
// Or with CommonJS:
//   var OpticsAgent = require('optics-agent');
exports.default = {
  configureAgent: _DefaultAgent.configureAgent,
  instrumentSchema: _DefaultAgent.instrumentSchema,
  middleware: _DefaultAgent.middleware,
  koaMiddleware: _DefaultAgent.koaMiddleware,
  instrumentHapiServer: _DefaultAgent.instrumentHapiServer,
  context: _DefaultAgent.context,
  Agent: _Agent2.default
};
exports.configureAgent = _DefaultAgent.configureAgent;
exports.instrumentSchema = _DefaultAgent.instrumentSchema;
exports.middleware = _DefaultAgent.middleware;
exports.koaMiddleware = _DefaultAgent.koaMiddleware;
exports.instrumentHapiServer = _DefaultAgent.instrumentHapiServer;
exports.context = _DefaultAgent.context;
exports.Agent = _Agent2.default;
